### DISCUSSION TITLE
How to declare Recursive Functions in cvc5?

### DISCUSSION THREAD

**Discussion by srod5125** [cvc5/cvc5#11199] [Q&A] [answered]
Hello, 

I am having trouble using recursive functions in cvc5. I am trying to declare the following function:
- f(x) = if x <= 0 then 0 else f( x - 1 )

Then I try to assert f(3) = 0. My issue is I keep getting unknown (incomplete) when checking for sat

```c++
#include <iostream>
#include <cvc5/cvc5.h>

int main()
{

    cvc5::TermManager tm = cvc5::TermManager();
    cvc5::Solver slv     = cvc5::Solver(tm);

    slv.setLogic("ALL");
    slv.setOption("produce-models", "true");


    auto x = tm.mkVar( tm.getIntegerSort() );

    auto fn = tm.mkConst(
        tm.mkFunctionSort( { tm.getIntegerSort() }, tm.getIntegerSort() )
    );

    // construct f(x) = if x <= 0 then 0 else f( x - 1 )
    auto rec_fn = slv.defineFunRec(
        fn ,
        { x } ,
        tm.mkTerm(
            cvc5::Kind::ITE,
            { tm.mkTerm(cvc5::Kind::LEQ,{ x, tm.mkInteger(0) }),
              tm.mkInteger(0),
              tm.mkTerm(cvc5::Kind::APPLY_UF, {
                  fn ,
                  tm.mkTerm(cvc5::Kind::SUB, { x, tm.mkInteger(1) })
              })
            }
        )
    );

    // adding or removing this assertion does not change the outcome
    slv.assertFormula(
      tm.mkTerm(
          cvc5::Kind::EQUAL,
          { rec_fn, fn }
      )
    );
    
    // f(3) 
    auto fn_3 = tm.mkTerm(
        cvc5::Kind::APPLY_UF,
        { rec_fn, tm.mkInteger(3) }
    );

    // assert f(3) == 0
    slv.assertFormula(
        tm.mkTerm(
            cvc5::Kind::EQUAL,
            { fn_3 , tm.mkInteger(0) }
        )
    );

    auto res = slv.checkSat();
    std::cout << "res: " << res << std::endl;
    // output: res: unknown (INCOMPLETE)
}
```
Any help would be greatly appreciated.
Thank you for your time

**Comment by ajreynol** [comment, marked as answer]
Sorry for the long delay in reply.  By default, when recursive functions are declared, cvc5 will generally not try to answer "sat".  You can use the experimental option `--fmf-fun` to make cvc5 try to answer "sat" in these cases, although this option is not model sound in the general case.


In detail, the issue is that in general one cannot assume that the definition of recursive functions do not introduce inconsistencies, e.g. think of the recursive definition:
f(x) = f(x)+1
This is non-terminating but by the semantics of SMT-LIB, any input with this definition is "unsat".

As mentioned, we do support an experimental option `--fmf-fun` which makes some assumptions that the recursively defined functions are consistent and uses finite model finding techniques for answer "sat" when recursive functions are present.  This is describe in the paper: https://homepage.divms.uiowa.edu/~ajreynol/ijcar16a.pdf
